JDP200203-16 Encja Order  testy

mackowit opened the pull request:
Inicjalna wersja klasy testowej dla encji Order

mackowit commented:
@Torayasu proszę o review. Dzieki!!

Torayasu commented on src/test/java/com/kodilla/ecommercee/domain/OrderTestSuite.java:
```diff
@@ -0,0 +1,173 @@
+        orderRepository.save(order);
+
+        //Then
+        Assert.assertEquals(1, orderRepository.count());
```
Oceniasz powodzenie testu, licząc ilość encji w repozytorium. Uzywając H2 do testów to jest spoko, ale jak baza zostanie podmieniona na wersję serwerowa przy dalszym życiu aplikacji, będziesz musiał reperować testy

mackowit replied:
pozostało, tak zostało przyjęte w poprzednich moich testach

Torayasu commented on src/test/java/com/kodilla/ecommercee/domain/OrderTestSuite.java:
```diff
@@ -0,0 +1,173 @@
+        try {
+            orderRepository.deleteAll();
+            productRepository.deleteAll();
+            userRepository.deleteAll();
```
Jak puścisz te testy na bazie produkcyjnej to wyrzucisz wszystkie rekordy z danych tabel. Możesz jako argumentu użyć tutaj encji która chcesz skasować, bądź kolekcji encji (wrzuć sobie np. produkty na listę).

hacker-msl replied:
o ile profil do testów jest ustawiony na H2 to co do zasady nie ma obawy takiej. Nie wiem czemu ktoś miałby odpalać testy na bazie produkcyjnej...

hacker-msl replied:
ja w celu zachowania idempotentności bym wyrzucił te sekcję clean-up do metody opatrzonej adnotacją @AfterEach (tak jest w Junit5) - generalnie chodzi o metodę, która odpala się po każdym teście.

Torayasu replied:
Hmm, jak komuś się niechcący zrobi Gradle Build i polecą od razu wszystkie testy jednostkowe, a  profil nie będzie zmieniony ? Nie będę się bardzo kłócił, ale wydaje mi się, że lepiej mieć testy, które zadziałają niezależnie od tego na jakiej bazie robimy. Ale może przesadzam :P

hacker-msl replied:
no i ja bym tego nie ładował do try-catch... wyjątki jakie lecą to są wyjątki niekontrolowane, więc program się skompiluje - to po pierwsze. Po drugie to zostawiona pusta sekcja catch spowoduje, żę jak coś się sypnie to... błędy pojawią się w innych testach prawdopodobnie niż test, w którym poleciał wyjątek. Ja to już bym wolał, żeby ten wyjątek sobie polecial w tym teście, w którym coś jest skopane. A może być skopane bo przy odpowiednim ustawieniu constrainów wystarczy, żę zmienisz kolejność wywołania metod deleteAll(), żeby clean-up poszedł w diabły.

hacker-msl replied:
@Torayasu jak ktoś podpina się z lokala do bazy produkcyjnej to jest idiotą. Poza tym i tak nie przejdą wtedy inne testy, które np. opierają się na metodach findAll() itp. Ja bym to zostawił.

mackowit replied:
sekcja //CleanUp przeniseiona do metody @After

Torayasu commented on src/test/java/com/kodilla/ecommercee/domain/OrderTestSuite.java:
```diff
@@ -0,0 +1,173 @@
+        try {
+            orderRepository.deleteAll();
+            userRepository.deleteAll();
+        } catch (Exception e) {
```
Jak wyżej

hacker-msl replied:
ja w celu zachowania idempotentności bym wyrzucił te sekcję clean-up do metody opatrzonej adnotacją @AfterEach (tak jest w Junit5) - generalnie chodzi o metodę, która odpala się po każdym teście.

hacker-msl replied:
no i ja bym tego nie ładował do try-catch... wyjątki jakie lecą to są wyjątki niekontrolowane, więc program się skompiluje - to po pierwsze. Po drugie to zostawiona pusta sekcja catch spowoduje, żę jak coś się sypnie to... błędy pojawią się w innych testach prawdopodobnie niż test, w którym poleciał wyjątek. Ja to już bym wolał, żeby ten wyjątek sobie polecial w tym teście, w którym coś jest skopane. A może być skopane bo przy odpowiednim ustawieniu constrainów wystarczy, żę zmienisz kolejność wywołania metod deleteAll(), żeby clean-up poszedł w diabły.

mackowit replied:
sekcja //CleanUp przeniseiona do metody @After

Torayasu commented on src/test/java/com/kodilla/ecommercee/domain/OrderTestSuite.java:
```diff
@@ -0,0 +1,173 @@
+        try {
+            orderRepository.deleteAll();
+            productRepository.deleteAll();
+            userRepository.deleteAll();
```
Jak wyżej

hacker-msl replied:
ja w celu zachowania idempotentności bym wyrzucił te sekcję clean-up do metody opatrzonej adnotacją @AfterEach (tak jest w Junit5) - generalnie chodzi o metodę, która odpala się po każdym teście.

hacker-msl replied:
no i ja bym tego nie ładował do try-catch... wyjątki jakie lecą to są wyjątki niekontrolowane, więc program się skompiluje - to po pierwsze. Po drugie to zostawiona pusta sekcja catch spowoduje, żę jak coś się sypnie to... błędy pojawią się w innych testach prawdopodobnie niż test, w którym poleciał wyjątek. Ja to już bym wolał, żeby ten wyjątek sobie polecial w tym teście, w którym coś jest skopane. A może być skopane bo przy odpowiednim ustawieniu constrainów wystarczy, żę zmienisz kolejność wywołania metod deleteAll(), żeby clean-up poszedł w diabły.

mackowit replied:
sekcja //CleanUp przeniseiona do metody @After

Torayasu commented on src/test/java/com/kodilla/ecommercee/domain/OrderTestSuite.java:
```diff
@@ -0,0 +1,173 @@
+        try {
+            orderRepository.deleteAll();
+            userRepository.deleteAll();
+        } catch (Exception e) {
```
Jak wyżej

hacker-msl replied:
ja w celu zachowania idempotentności bym wyrzucił te sekcję clean-up do metody opatrzonej adnotacją @AfterEach (tak jest w Junit5) - generalnie chodzi o metodę, która odpala się po każdym teście. 

no i ja bym tego nie ładował do try-catch... wyjątki jakie lecą to są wyjątki niekontrolowane, więc program się skompiluje - to po pierwsze. Po drugie to zostawiona pusta sekcja catch spowoduje, żę jak coś się sypnie to... błędy pojawią się w innych testach prawdopodobnie niż test, w którym poleciał wyjątek. Ja to już bym wolał, żeby ten wyjątek sobie polecial w tym teście, w którym coś jest skopane. A może być skopane bo przy odpowiednim ustawieniu constrainów wystarczy, żę zmienisz kolejność wywołania metod deleteAll(), żeby clean-up poszedł w diabły.

mackowit replied:
sekcja //CleanUp przeniseiona do metody @After

hacker-msl commented:
@Torayasu czy Ty zatwierdziłeś ten PR? wolę dopytać bo napisałeś "gra" ale status nie jest jako "approved" u Ciebie...

mackowit commented:
@Torayasu @hacker-msl poprawione testy, usunięte try/catch, dodane before/after, dodatkowo dostosowane do zmienionej encji Order (zamiast listy produktów jest mapa), proszę o review

hacker-msl commented on src/main/java/com/kodilla/ecommercee/domain/Order.java:
```diff
@@ -14,30 +16,37 @@
+        this.id = id;
+        this.products = new HashMap<>(products);
+        this.user = user;
+        this.orderDate = LocalDate.now();
```
w takich sytuacjach zdecydowanie wolę napisać 1 prywatny konstruktor i kilka nazwanych, statycznych metod, które mają za zadanie tworzyć dany obiekt. 

public static ofCart(Cart cart) {
  return new Order(... );
}

public static of(... ) {
  return new Order( ... ) ;
}

to tak na przyszłość :-)

hacker-msl commented on src/test/java/com/kodilla/ecommercee/domain/OrderTestSuite.java:
```diff
@@ -0,0 +1,112 @@
+        orderRepository.save(order1);
+
+        //Then
+        Assert.assertEquals(3, orderRepository.findById(order1.getId()).get().getProducts().size());
```
może statyczny import byłby lepszy  ? to tak na przyszłość

hacker-msl commented on src/test/java/com/kodilla/ecommercee/domain/OrderTestSuite.java:
```diff
@@ -0,0 +1,112 @@
+        //Given
+        //When
+        //Then
+        Assert.assertEquals(3, orderRepository.count());
```
może statyczny import byłby lepszy  ? to tak na przyszłość

mackowit commented:
@hacker-msl @Sebeni Teraz przepraszam za głupie pytania ale nie chcę znowu narobić dziadostwa.
Ściągnąłem sobie pull i oczywiście były dwa konflikty z ww plikami, zaakceptowałem w intellij zmiany "theirs" ale mam teraz w git status całą masę plików na żółto do commitowania, proszę Was teraz o poradę jak to zrobić żeby wypchnąć teraz zmianę tylko w tych dwóch plikach? Dzięki z góry za podpowiedź, będę już wiedział na przyszłość

mackowit commented:
C:\Users\Samsung\Projects\project-jdp-2002-03>git status
On branch JDP200203-16
Your branch is up to date with 'origin/JDP200203-16'.

All conflicts fixed but you are still merging.
  (use "git commit" to conclude merge)

Changes to be committed:
        renamed:    src/main/java/com/kodilla/ecommercee/OrderController.java -> src/main/java/com/kodilla/ecommercee/controller/OrderController.java
        modified:   src/main/java/com/kodilla/ecommercee/domain/Cart.java
        modified:   src/main/java/com/kodilla/ecommercee/domain/CartDto.java
        modified:   src/main/java/com/kodilla/ecommercee/domain/Group.java
        modified:   src/main/java/com/kodilla/ecommercee/domain/Order.java
        modified:   src/main/java/com/kodilla/ecommercee/domain/OrderDto.java
        modified:   src/main/java/com/kodilla/ecommercee/domain/Product.java
        new file:   src/main/java/com/kodilla/ecommercee/exception/OrderNotFoundException.java
        deleted:    src/main/java/com/kodilla/ecommercee/exceptions/ProductNotFoundException.java
        modified:   src/main/java/com/kodilla/ecommercee/mapper/GroupMapper.java
        new file:   src/main/java/com/kodilla/ecommercee/mapper/OrderMapper.java
        modified:   src/main/java/com/kodilla/ecommercee/repository/OrderRepository.java
        modified:   src/main/java/com/kodilla/ecommercee/service/CartService.java
        modified:   src/main/java/com/kodilla/ecommercee/service/OrderService.java
        modified:   src/test/java/com/kodilla/ecommercee/controller/GroupControllerTest.java
        new file:   src/test/java/com/kodilla/ecommercee/controller/OrderControllerTest.java

Sebeni commented:
@mackowit te wszystkie pliki, o których piszesz, że są na żółto, to pliki, które zostały zmergowane bez konfliktów (nie zawsze merge tworzy konflikt). Jako że są inne od Twoich w lokalnym repo to Ci się pojawiają. Imo możesz je wszystkie dodać i pushować, bo nie ma w nich konfliktów z masterem (są teraz identyczne na Twoim branchu i na masterze).

mackowit commented:
i wypchnąć te wszystkie pliki do repozytorium?

Sebeni commented:
tak

Sebeni commented:
one w repo już i tak są (na masterze). Po prostu "synchronizujesz" swojego brancha z masterem.

mackowit commented:
ok, dzięki wielkie, poszło i jest ok:)

mackowit commented:
@hacker-msl @Torayasu konflikty usunięte, gotowe do zatiwerdzenia
